File: FastCP.rs
fn get_tuple_at(mut index: usize, sets: &[Vec<i32>]) -> Vec<i32> {
    let mut tuple = vec!(0; sets.len());
    for i in (0..sets.len()).rev() {
        tuple[i] = sets[i][index % sets[i].len()];
        index /= sets[i].len();
    }
    tuple
}

fn main() {
    let mut sets_vec = vec![vec![0i32; 5]; 3];
    let mut sets_arr = [[0i32; 5]; 3];
    for i in 0..5 {
        for j in 0..3 {
            sets_vec[j][i] = i as i32 + 1;
            sets_arr[j][i] = i as i32 + 1;
        }
    }
    println!("{:?}", get_tuple_at(2usize, &sets_vec));
}

File: FastCP.hs
module FastCP
  (
    getNthTuple
  ) where

{-
Recursively go through all the sets and extract the (n `mod` length rev) element

Modify n to be (n `div` length rev) each recursion
-}
getNthTuple' :: Int -> [[a]] -> [a]
getNthTuple' _ [] = []
getNthTuple' n (rev:revs) = rev !! currIndx : getNthTuple' newN revs
    where 
        lenrev = length rev
        currIndx = n `mod` lenrev
        newN = n `div` lenrev

{-
Invoke getNthTuple' and pass in a reversed list of sets, as well as the index

Reverse the final result and return it
-}
getNthTuple :: Int -> [[a]] -> [a]
getNthTuple = (reverse .) . (. reverse) . getNthTuple'
